### PR TITLE
refactor(standards): Relaton fetcher API cleanup (#123, #124, #125)

### LINKED ISSUE #123: feat(standards): implement RelatonLiveFetcher.search() for ISO/IEC
Context: PR #122 registered `RelatonLiveFetcher` under the `ISO`, `IEC`, and `ISO/IEC` keys in `StandardsClient._fetchers`, but the fetcher only exposes `fetch(identifier)` — no `search(query, *, limit)` method. `StandardsClient.search()` guards with `hasattr(f, "search")`, so search queries for those bodies silently return `[]`.

This is documented in `StandardsClient.search`'s docstring (which still lists only NIST/IETF/W3C/ETSI as valid body filters), so the contract is honest. But it's a visible asymmetry: users can `get` an ISO/IEC standard by identifier but can't search the Relaton corpus.

### Scope

- Add `async def search(self, query: str, *, limit: int = 10) -> list[StandardRecord]` to `RelatonLiveFetcher` (or a companion searcher class that wraps the synced cache).
- Decision needed: search the local synced cache (requires `sync-standards` to have run), search the upstream GitHub via the search API, or both (cache-first, fallback to live).
- Update `StandardsClient.search` docstring to include `ISO`, `IEC`, `ISO/IEC` once implemented.
- Add integration tests covering body-filtered and all-bodies search paths.

### Deferred from

PR #122 final review — the `hasattr(f, "search")` filter was added specifically to accommodate this deferral.

### PR DESCRIPTION
**Pull request by pvliesdonk**
## Summary

- **#125** — Rename `RelatonLiveFetcher.fetch` → `.get` for API symmetry with all other Tier 1 fetchers
- **#123** — Add `search_synced_standards()` to `ScholarCache` + `CacheProtocol`; implement `RelatonLiveFetcher.search()` backed by the synced cache; wire real cache through `StandardsClient` and `_server_deps`
- **#124** — Replace `hasattr`-based fetcher dispatch with a typed `_StandardsFetcher` Protocol (`@runtime_checkable`); remove all `cast()` and `hasattr()` calls; `_fetchers` dict is now `dict[str, _StandardsFetcher]`
- **Fix** — Escape `%` and `_` LIKE wildcards in `search_synced_standards` query to prevent semantic wildcard injection

Closes #123, #124, #125.

## Test Plan

- [ ] `uv run pytest -x -q` — 884 tests pass
- [ ] `uv run ruff check --fix . && uv run ruff format . && uv run ruff format --check .` — clean
- [ ] `uv run mypy src/` — no errors (46 source files)
- [ ] New tests: `search_synced_standards` (identifier match, title match, live-fetched exclusion, source filter, limit, wildcard escaping)
- [ ] New tests: `RelatonLiveFetcher.search()` with and without cache
- [ ] New test: all registered fetchers satisfy `_StandardsFetcher` Protocol at runtime

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/scholar_mcp/_relaton_live.py [hunk 1]
```diff
@@ -102,7 +109,26 @@ async def _try_repo(self, repo: str, slug: str) -> dict[str, object] | None:
...
+        Args:
+            query: Substring to match against identifier and title.
+            limit: Maximum number of results.
+
+        Returns:
+            List of matching ``StandardRecord`` dicts.
+        """
+        if self._cache is None:
+            return []
+        return await self._cache.search_synced_standards(query, limit=limit)
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/scholar_mcp/_relaton_live.py]: Declining this one. The design decision for #123 was explicitly to make `search()` part of the symmetric `_StandardsFetcher` Protocol (alongside `.get()`) and let each fetcher own its own data source. `RelatonLiveFetcher` uses the synced cache as its data source because Relaton has no live search API — the fetcher isn't "cache-aware," it's "my source of truth is the synced DB." It receives any `CacheProtocol` implementation via DI, so it's decoupled from `ScholarCache` specifically.

Moving the cache integration up to the tool layer would fracture the Protocol (tools would need to know which fetchers are cache-backed vs. API-backed) and duplicate per-body dispatch logic that `StandardsClient` already owns. The current placement keeps the tool layer uniform.

[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/scholar_mcp/_relaton_live.py]: Fixed in abb87a4. Went with the "three distinct instances" option: `RelatonLiveFetcher.__init__` now accepts `source`, and `StandardsClient` registers `source="ISO"`, `source="IEC"`, and `source=None` for the ISO/IEC key. The no-body search path now deduplicates merged results by identifier (since the ISO/IEC variant is unfiltered and overlaps with the ISO/IEC ones). The no-body `get()` path deduplicates by fetcher type so the slug lookup only runs once.

File: src/scholar_mcp/_relaton_live.py [hunk 2]
```diff
@@ -102,7 +109,26 @@ async def _try_repo(self, repo: str, slug: str) -> dict[str, object] | None:
...
+        context without a cache).
+
+        Args:
+            query: Substring to match against identifier and title.
+            limit: Maximum number of results.
+
+        Returns:
+            List of matching ``StandardRecord`` dicts.
+        """
+        if self._cache is None:
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/scholar_mcp/_relaton_live.py:142]: Added in abb87a4.

File: src/scholar_mcp/_relaton_live.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/scholar_mcp/_relaton_live.py]: Fixed in abb87a4 — see the detailed response on the earlier thread for the same issue. Three distinct `RelatonLiveFetcher` instances per body key, deduped-by-identifier in the all-bodies search path, deduped-by-type in the all-bodies get path.

File: tests/test_standards_client.py [hunk 3]
```diff
@@ -1206,3 +1206,43 @@ async def test_standards_client_get_fallback_to_fetchers() -> None:
...
+        "body": "ISO",
+        "status": "published",
+        "full_text_available": False,
+    }
+    mock_cache = AsyncMock()
+    mock_cache.search_synced_standards = AsyncMock(return_value=[record])
+
+    async with httpx.AsyncClient() as http:
+        client = StandardsClient(http, cache=mock_cache)
+        results = await client.search("9001", body="ISO")
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply tests/test_standards_client.py:1231]: Strengthened in abb87a4 — now asserts `search_synced_standards` was awaited with `source="ISO", limit=10`. Also added symmetric tests for `body="IEC"` (asserting `source="IEC"`) and `body="ISO/IEC"` (asserting `source=None`) as regression guards.

File: src/scholar_mcp/_cache.py [hunk 4]
```diff
@@ -1126,6 +1126,56 @@ async def list_synced_standard_ids(self, source: str) -> set[str]:
...
+                ``None`` to search all synced bodies.
+            limit: Maximum number of results.
+
+        Returns:
+            List of matching ``StandardRecord`` dicts.
+        """
+        db = _require_open(self._db)
+        escaped = query.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
+        pattern = f"%{escaped}%"
+        if source is not None:
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/scholar_mcp/_cache.py:1158]: Unified in abb87a4.

File: src/scholar_mcp/_standards_client.py [hunk 5]
```diff
@@ -1024,37 +1051,34 @@ async def search(
...
+        # Search all sources concurrently. The three RelatonLiveFetcher
+        # instances may return overlapping records (the ISO/IEC variant with
+        # source=None is unfiltered), so dedup merged results by identifier.
         results_per_body = await asyncio.gather(
-            *(
-                f.search(query, limit=limit)
-                for f in self._fetchers.values()
-                if hasattr(f, "search")
-            ),
+            *(f.search(query, limit=limit) for f in self._fetchers.values()),
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/scholar_mcp/_standards_client.py]: Fixed in 79109ae. Extracted `_one_fetcher_per_type()` helper that deduplicates by concrete fetcher class with last-wins overwrite. For `RelatonLiveFetcher` the insertion order `ISO → IEC → ISO/IEC` ensures the unfiltered `source=None` variant is the one kept, so the no-body search path now fires a single Relaton cache query instead of three. Added a regression test (`test_standards_client_search_no_body_hits_relaton_cache_once`) that asserts `await_count == 1`.

File: src/scholar_mcp/_standards_client.py [hunk 6]
```diff
@@ -1073,18 +1097,16 @@ async def get(self, identifier: str) -> StandardRecord | None:
...
-            else:
-                result = cast("StandardRecord | None", await fetcher.get(identifier))
+                return await fetcher.get(canonical)
+
+        # No local resolution — try each fetcher (deduped by type: the three
+        # RelatonLiveFetcher instances do identical slug-based lookup for
+        # get(), so call only one of them).
+        seen_types: dict[type[object], _StandardsFetcher] = {}
+        for f in self._fetchers.values():
+            seen_types.setdefault(type(f), f)
```
[the remark this answers is not in the document]
Comment (pvliesdonk) [reply src/scholar_mcp/_standards_client.py]: Documented in 79109ae. The dedup now lives in `_one_fetcher_per_type()` with a docstring that spells out the deliberate contract: last-wins overwrite relying on insertion order, intentional for `RelatonLiveFetcher` to keep the `source=None` variant, and a note that any future second instance of a non-Relaton type would replace the earlier one (register a distinct subclass if that ever matters).
